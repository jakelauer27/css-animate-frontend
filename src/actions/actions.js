export const signIn = (user) => ({
  type: 'SIGN_IN',
  user
})

export const signOut = () => ({
  type: 'SIGN_OUT'
})

export const loadPrebuiltAnimations = (prebuiltAnimations) => ({
  type: 'LOAD_PREBUILT_ANIMATIONS',
  prebuiltAnimations
})

export const updateCurrentAnimation = (animation) => ({
  type: 'UPDATE_CURRENT_ANIMATION',
  animation
})

export const saveOriginalAnimation = (animation) => ({
  type: 'SAVE_ORIGINAL_ANIMATION',
  animation
})

export const loadMyAnimations = (myAnimations) => ({
  type: 'LOAD_MY_ANIMATIONS',
  myAnimations
})

export const deleteMyAnimation = (id) => ({
  type: 'DELETE_MY_ANIMATION',
  id
})

export const selectAnimationForEdit = (animation) => ({
  type: 'SELECT_ANIMATION_FOR_EDIT',
  animation
})

export const removeAnimationForEdit = () => ({
  type: 'REMOVE_ANIMATION_FOR_EDIT'
})


