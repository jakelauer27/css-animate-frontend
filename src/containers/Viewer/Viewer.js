import React, { Component } from 'react'
import { connect } from 'react-redux'
import { PropTypes } from 'prop-types'

export class Viewer extends Component {
  constructor() {
    super()
    this.state = {
      animation: {},
      play: true
    }
  }
  
  playAnimation() {
    const { currentAnimation } = this.props
    this.setState({
      animation: {
        animation: `${currentAnimation.name} ${currentAnimation.duration} ${currentAnimation.timingFunction} ${currentAnimation.delay} ${currentAnimation.iterationCount} ${currentAnimation.direction}`,
        animationFillMode: `${currentAnimation.fillMode}`
      },
      play: false
    })
  }

  resetAnimation(waitTime) {
    setTimeout( 
      this.setState({
        animation: {},
        play: true
      })
    , waitTime)
  }

  render() {
    const { currentAnimation } = this.props
    if  (!currentAnimation) {
      return <div></div>
    }
    return (
      <div className='viewer-component'>
        <h2 className='current-animation-label'>*{currentAnimation.name}*</h2>
        <div className='viewer-container'>
          <div className='viewer'>
            <div className='square square-0'
                  style={this.state.animation}
                  onAnimationEnd={() => this.resetAnimation(1000)}>
                <h4 className='main-title'>ani<span>Mate</span></h4>
            </div>
          </div>
        </div>
        <div className='editor-bottom-btns-container'>
          <button className={`lower-btn play-btn`} 
            onClick={() => this.playAnimation()}
            disabled={!this.state.play}>play</button>
          <button className={`lower-btn stop-btn`} 
            onClick={() => this.resetAnimation(0)}
            disabled={this.state.play}>stop</button>
        </div>
      </div>
    )
  }
}

export const mapStateToProps = (state) => ({
  currentAnimation: state.currentAnimation.properties
})

Viewer.propTypes = {
  currentAnimation: PropTypes.object
}

export default connect(mapStateToProps)(Viewer)