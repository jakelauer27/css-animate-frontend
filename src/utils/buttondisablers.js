export const prebuiltButtonDisabler = (input) => {
  if(!input) {
    document.querySelector('.play-btn').setAttribute('disabled', true)
    document.querySelector('.save-as-btn').setAttribute('disabled', true)
  } else {
    document.querySelector('.play-btn').removeAttribute('disabled')
    document.querySelector('.save-as-btn').removeAttribute('disabled')
  }
}
export const buttonDisabler = (input) => {
  if(!input) {
    document.querySelector('.save-as-btn').setAttribute('disabled', true)
    document.querySelector('.save-btn').setAttribute('disabled', true)
    document.querySelector('.play-btn').setAttribute('disabled', true)
  } else {
    document.querySelector('.save-btn').removeAttribute('disabled')
    document.querySelector('.save-as-btn').removeAttribute('disabled')
    document.querySelector('.play-btn').removeAttribute('disabled')
  }
}