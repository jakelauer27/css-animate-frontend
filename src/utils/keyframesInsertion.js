

export const updateKeyframes = (newKeyframes) => {
  const sheet = document.styleSheets[document.styleSheets.length - 1];
  let ruleKeys = Object.keys(sheet.cssRules)
  let keyframeToDeleteIndex = ruleKeys.find(rule => {
    return sheet.cssRules[rule].name === newKeyframes.name;
  })
  sheet.deleteRule(keyframeToDeleteIndex);
  sheet.insertRule(`@keyframes ${newKeyframes.name} 
    ${stringifyKeyframes(newKeyframes)}`, sheet.cssRules.length);
}

export const stringifyKeyframes = (keyframe) => {
  const stringified = keyframe.sections.reduce( (stages, section, i) => {
    let sectionProps = section.properties.reduce( (propsObj, prop, i) => {
      if (i === section.properties.length - 1) {
        return `${propsObj} ${prop.name}: ${prop.value};}`
      }
      return `${propsObj} ${prop.name}: ${prop.value};` ;
    }, `{`)
    if (i === keyframe.sections.length - 1) {
      return stages + section.label + sectionProps + "}"
    }
    return stages + section.label + sectionProps
  }, "{")
  return stringified
}