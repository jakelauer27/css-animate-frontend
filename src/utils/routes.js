export const routes = [`/properties/selectAnimation`, `/properties/signup`, `/properties/login`, `/properties/howto/general`, `/keyframes`, `/properties`, '/', `/properties/howto/properties`,`/properties/howto/keyframes`, '/properties/selectAnimation/create', `/properties/copy`, '/error']