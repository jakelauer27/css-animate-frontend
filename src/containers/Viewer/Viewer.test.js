import { loadAnimation } from '../../actions/actions'
import * as CSSInsertion from '../../utils/keyframesInsertion'
import React from 'react'
import { Viewer, mapStateToProps, mapDispatchToProps } from './Viewer'
import { shallow } from 'enzyme'

jest.mock('../../utils/keyframesInsertion')

describe('Viewer', () => {
  let wrapper
  const mockAnimation = {
      'name': 'slideInX',
      'duration': '1.5s',
      'timingFunction': 'ease',
      'delay': '0s',
      'iterationCount': '1',
      'direction': 'normal',
      'fillMode': 'forwards'
  }

  beforeEach(() => {
    wrapper = shallow(<Viewer currentAnimation={mockAnimation} />)
  })

  it('Should match the snapshot' , () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('Should match the snapshot if no animation' , () => {
    wrapper = shallow(<Viewer currentAnimation={null} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('the play button should trigger playAnimation method', () => {
    const spy = jest.spyOn(wrapper.instance(), 'playAnimation')
    wrapper.find('.play-btn').simulate('click')
    expect(spy).toHaveBeenCalled()
  })

  it('the reset button should trigger resetAnimation method', () => {
    const spy = jest.spyOn(wrapper.instance(), 'resetAnimation')
    wrapper.find('.stop-btn').simulate('click')
    expect(spy).toHaveBeenCalledWith(0)
  })

  it('animationEnd should trigger resetAnimation method', () => {
    const spy = jest.spyOn(wrapper.instance(), 'resetAnimation')
    wrapper.find('.square').simulate('animationEnd')
    expect(spy).toHaveBeenCalledWith(1000)
  })

  describe('playAnimation', () => {

    it('should set play to false in state', () => {
      const expected = false
      wrapper.instance().playAnimation()
      expect(wrapper.state().play).toEqual(expected)
    })

    it('should set animation to the short hand version of the passed in animation', () => {
      const expected = {"animation": "slideInX 1.5s ease 0s 1 normal", "animationFillMode": "forwards"}
      wrapper.instance().playAnimation()
      expect(wrapper.state().animation).toEqual(expected)
    })
  })

  describe('resetAnimation', () => {

    jest.useFakeTimers();
    
    it('should set play to true after waiting the specified time', () => {
      const expected = true
      wrapper.instance().resetAnimation(1)
      expect(wrapper.state().play).toEqual(expected)
    })

    it('should call setTimeout with this.setState', () => {
      wrapper.instance().resetAnimation(1)
      expect(setTimeout).toHaveBeenCalled()
      expect(setTimeout).toHaveBeenCalledWith(wrapper.instance().setState(), 1)
    })

    it('should set animation to an empty object after waiting the specified time', () => {
      const expected = {}
      wrapper.instance().resetAnimation(1)
      expect(wrapper.state().animation).toEqual(expected)
    })
  })

  describe('mapStateToProps', () => {

    it('should return a props object with an animation property', () => {
      const mockState = {currentAnimation: mockAnimation}
      const expected = {currentAnimation: mockAnimation.properties}
      expect(mapStateToProps(mockState)).toEqual(expected)
    })
  })

})