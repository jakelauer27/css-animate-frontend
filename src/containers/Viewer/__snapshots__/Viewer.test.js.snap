// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Viewer Should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Viewer
    currentAnimation={
      Object {
        "delay": "0s",
        "direction": "normal",
        "duration": "1.5s",
        "fillMode": "forwards",
        "iterationCount": "1",
        "name": "slideInX",
        "timingFunction": "ease",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2
          className="current-animation-label"
        >
          *
          slideInX
          *
        </h2>,
        <div
          className="viewer-container"
        >
          <div
            className="viewer"
          >
            <div
              className="square square-0"
              onAnimationEnd={[Function]}
              style={Object {}}
            >
              <h4
                className="main-title"
              >
                ani
                <span>
                  Mate
                </span>
              </h4>
            </div>
          </div>
        </div>,
        <div
          className="editor-bottom-btns-container"
        >
          <button
            className="lower-btn play-btn"
            disabled={false}
            onClick={[Function]}
          >
            play
          </button>
          <button
            className="lower-btn stop-btn"
            disabled={true}
            onClick={[Function]}
          >
            stop
          </button>
        </div>,
      ],
      "className": "viewer-component",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "*",
            "slideInX",
            "*",
          ],
          "className": "current-animation-label",
        },
        "ref": null,
        "rendered": Array [
          "*",
          "slideInX",
          "*",
        ],
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="viewer"
          >
            <div
              className="square square-0"
              onAnimationEnd={[Function]}
              style={Object {}}
            >
              <h4
                className="main-title"
              >
                ani
                <span>
                  Mate
                </span>
              </h4>
            </div>
          </div>,
          "className": "viewer-container",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="square square-0"
              onAnimationEnd={[Function]}
              style={Object {}}
            >
              <h4
                className="main-title"
              >
                ani
                <span>
                  Mate
                </span>
              </h4>
            </div>,
            "className": "viewer",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <h4
                className="main-title"
              >
                ani
                <span>
                  Mate
                </span>
              </h4>,
              "className": "square square-0",
              "onAnimationEnd": [Function],
              "style": Object {},
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "ani",
                  <span>
                    Mate
                  </span>,
                ],
                "className": "main-title",
              },
              "ref": null,
              "rendered": Array [
                "ani",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Mate",
                  },
                  "ref": null,
                  "rendered": "Mate",
                  "type": "span",
                },
              ],
              "type": "h4",
            },
            "type": "div",
          },
          "type": "div",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <button
              className="lower-btn play-btn"
              disabled={false}
              onClick={[Function]}
            >
              play
            </button>,
            <button
              className="lower-btn stop-btn"
              disabled={true}
              onClick={[Function]}
            >
              stop
            </button>,
          ],
          "className": "editor-bottom-btns-container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "play",
              "className": "lower-btn play-btn",
              "disabled": false,
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "play",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "stop",
              "className": "lower-btn stop-btn",
              "disabled": true,
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "stop",
            "type": "button",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2
            className="current-animation-label"
          >
            *
            slideInX
            *
          </h2>,
          <div
            className="viewer-container"
          >
            <div
              className="viewer"
            >
              <div
                className="square square-0"
                onAnimationEnd={[Function]}
                style={Object {}}
              >
                <h4
                  className="main-title"
                >
                  ani
                  <span>
                    Mate
                  </span>
                </h4>
              </div>
            </div>
          </div>,
          <div
            className="editor-bottom-btns-container"
          >
            <button
              className="lower-btn play-btn"
              disabled={false}
              onClick={[Function]}
            >
              play
            </button>
            <button
              className="lower-btn stop-btn"
              disabled={true}
              onClick={[Function]}
            >
              stop
            </button>
          </div>,
        ],
        "className": "viewer-component",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "*",
              "slideInX",
              "*",
            ],
            "className": "current-animation-label",
          },
          "ref": null,
          "rendered": Array [
            "*",
            "slideInX",
            "*",
          ],
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="viewer"
            >
              <div
                className="square square-0"
                onAnimationEnd={[Function]}
                style={Object {}}
              >
                <h4
                  className="main-title"
                >
                  ani
                  <span>
                    Mate
                  </span>
                </h4>
              </div>
            </div>,
            "className": "viewer-container",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <div
                className="square square-0"
                onAnimationEnd={[Function]}
                style={Object {}}
              >
                <h4
                  className="main-title"
                >
                  ani
                  <span>
                    Mate
                  </span>
                </h4>
              </div>,
              "className": "viewer",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <h4
                  className="main-title"
                >
                  ani
                  <span>
                    Mate
                  </span>
                </h4>,
                "className": "square square-0",
                "onAnimationEnd": [Function],
                "style": Object {},
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "ani",
                    <span>
                      Mate
                    </span>,
                  ],
                  "className": "main-title",
                },
                "ref": null,
                "rendered": Array [
                  "ani",
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "Mate",
                    },
                    "ref": null,
                    "rendered": "Mate",
                    "type": "span",
                  },
                ],
                "type": "h4",
              },
              "type": "div",
            },
            "type": "div",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <button
                className="lower-btn play-btn"
                disabled={false}
                onClick={[Function]}
              >
                play
              </button>,
              <button
                className="lower-btn stop-btn"
                disabled={true}
                onClick={[Function]}
              >
                stop
              </button>,
            ],
            "className": "editor-bottom-btns-container",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "play",
                "className": "lower-btn play-btn",
                "disabled": false,
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "play",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "stop",
                "className": "lower-btn stop-btn",
                "disabled": true,
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "stop",
              "type": "button",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`Viewer Should match the snapshot if no animation 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Viewer
    currentAnimation={null}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {},
    "ref": null,
    "rendered": null,
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
